Objects/typeobject.c

```diff
@@ -6691,49 +6674,84 @@ object_set_class(PyObject *self, PyObject *value, void *closure)
         /* Changing the class will change the implicit dict keys,
          * so we must materialize the dictionary first. */
         if (oldto->tp_flags & Py_TPFLAGS_INLINE_VALUES) {
-            PyDictObject *dict = _PyObject_MaterializeManagedDict(self);
+            PyDictObject *dict = _PyObject_GetManagedDict(self);
             if (dict == NULL) {
-                return -1;
+                dict = _PyObject_MaterializeManagedDict_LockHeld(self);
+                if (dict == NULL) {
+                    return -1;
+                }
             }
 
-            bool error = false;
-
-            Py_BEGIN_CRITICAL_SECTION2(self, dict);
-
             // If we raced after materialization and replaced the dict
             // then the materialized dict should no longer have the
             // inline values in which case detach is a nop.
+            // Note: we don't need to lock here because the world should be stopped.
+
             assert(_PyObject_GetManagedDict(self) == dict ||
                    dict->ma_values != _PyObject_InlineValues(self));
```

Comment:
This can be simplified now:

```suggestion
            assert(_PyObject_GetManagedDict(self) == dict);
```